certora/specs/StakingRewardStreams.spec

```diff
@@ -0,0 +1,283 @@
+    ) {
+        preserved {
+            requireInvariant epochSumsLETotalRegistered();
+        }
+    }
+
+
+/// @title Reward amount per epoch is not greater than total registered reward
+invariant rewardAmountLessThanTotal(address rewarded, address reward, uint48 epoch)
+    rewardAmount(rewarded, reward, epoch) <= totalRewardRegistered(rewarded, reward);
+
+
+
+/*******  High Level Properties  *******/
+
+
+/** @title Invariant: reward token solvency.
+The expected balance of rewarded and reward token:
+reward.balanceOf(this) === forall rewarded : sum of (totalRegistered - totalClaimed ) 
+rewarded.balanceOf(this) === sum forall account balanceOf(account,rewarded);
+
+The same token can be both a reward and rewarded. 
+*/
+
+invariant solvency(address token) 
+    to_mathint(externalBalanceOf(token, currentContract)) >=  sumBalancesPerRewarded[token] + sumOfNotDistributed[token]
+
+    {
+        preserved ERC20Caller.externalTransfer(address erc20, address to, uint256 amount) with (env e) {
+            //assume ( better to prove this) that current contract does not have a dynamic call 
+            require e.msg.sender != currentContract;
+        }
+    }
+    
+
+
+/*******  Staking Properties  *******/
+
+/// @title Staking increases staked balance by given amount
+rule stakeIntegrity(address rewarded, uint256 amount) {
+    env e;
+    require e.msg.sender != currentContract.evc;
+    uint256 preBalance = balanceOf(e.msg.sender, rewarded);
+
+    stake(e, rewarded, amount);
+
+    uint256 postBalance = balanceOf(e.msg.sender, rewarded);
+
+    assert (
+        amount != max_uint256 => to_mathint(postBalance) == preBalance + amount,
+        "Staking increases staked balance by given amount"
+    );
+}
+
+
+/*******  Rewards Properties  *******/
+/// @title An example showing rewards can be given
+rule canBeRewarded(address rewarded, address reward, address recipient, bool forfeitRecentReward) {
+    uint256 preBalance = externalBalanceOf(rewarded, recipient);
+
+    env e;
+    claimReward(e, rewarded, reward, recipient, forfeitRecentReward);
+
+    uint256 postBalance = externalBalanceOf(rewarded, recipient);
+    satisfy postBalance > preBalance;
+}
+
+
+/// @title Those that stake more should earn more rewards
+rule stakeMoreEarnMore(
+    address staker1,
+    address staker2,
+    address rewarded,
+    address reward,
+    bool forfeitRecentReward
+) {
+    env e1;
+    env e2;
+    
+
+    uint256 earned1 = earnedReward(e1, staker1, rewarded, reward, forfeitRecentReward);
+    uint256 earned2 = earnedReward(e2, staker1, rewarded, reward, forfeitRecentReward);
+
+    require e2.block.timestamp > e1.block.timestamp;
+    uint256 earned1Late = earnedReward(e1, staker1, rewarded, reward, forfeitRecentReward);
+    uint256 earned2Late = earnedReward(e2, staker1, rewarded, reward, forfeitRecentReward);
+
+    mathint diff1 = earned1Late - earned1;
```

Comment:
it seems that `earned1Late == earned1` && `earned2Late == earned2` which implies `diff1 
 == diff2 == 0`. I guess it should be done as follows instead:
 
```
uint256 earned1 = earnedReward(e1, staker1, rewarded, reward, forfeitRecentReward);
uint256 earned2 = earnedReward(e2, staker1, rewarded, reward, forfeitRecentReward);

require e2.block.timestamp > e1.block.timestamp;
uint256 earned1Late = earnedReward(e2, staker1, rewarded, reward, forfeitRecentReward);
uint256 earned2Late = earnedReward(e2, staker2, rewarded, reward, forfeitRecentReward);
 ```

Reply:
yes agree it was wrong. Proving the assertion requires first proving some other properties on the reward calculation. So change for now to only satisfy